inductiva/benchmarks/methods.py

```diff
@@ -0,0 +1,348 @@
+    """
+
+    print()
+    print(__("tasks_ran"))
+    for key, value in already_ran.items():
+        print(f"--->{key}: {len(value)}")
+    print()
+    print(__("tasks_to_run"))
+    for key, value in tasks_to_run.items():
+        print(f"--->{key}: {value}")
+    print()
+
+
+def _run(project: Project,
+         input_files: str,
+         replicas: int,
+         machines: List[Dict[str, Any]],
+         simulator: Simulator,
+         machine_class: BaseMachineGroup = None,
+         input_args: Dict[str, Union[Callable, Any]] = None,
+         machine_args: Dict[str, Union[Callable, Any]] = None,
+         simulator_args: Dict[str, Union[Callable, Any]] = None):
+    """Runs a benchmark.
+    This method runs a benchmark using the provided parameters.
+    Args:
+        project (Project): Project to run the benchmark on.
+    """
+    # create dict with resource type: list of tasks ran on that resource
+    # Example
+    # { "n1-standard-4": [task1, task2], "n1-standard-8": [task3]}
+    current_project_tasks = _tasks_by_vm_type(project)
+    to_run = _compute_tasks_to_run(machines, current_project_tasks, replicas)
+
+    print()
+    if not to_run:
+        print(__("all_tasks_run"))
+        project.close()
+        return
+
+    _print_info(current_project_tasks, to_run)
+
+    for machine, replicas_to_run in to_run.items():
+
+        machine_args_current = _replace_callable_from_dict(
+            machine_args, machine)
+
+        print(f"\nCreating machine group {machine} with {machine_args_current}")
+
+        # Use MachineGroup as default value for machine_class
+        machine_class = machine_class or MachineGroup
+
+        resource = machine_class(machine_type=machine,
+                                 max_idle_time=datetime.timedelta(minutes=1),
+                                 **machine_args_current)
+
+        print(f"\nRunning {simulator}")
+        if resource is not None:
+            while not _can_start_resource(resource):
+                print("This machine will exceed the current quotas.\n"
+                      "Going to sleep and trying again later.")
+                time.sleep(60)
+
+            resource.start()
+
+        for replica in range(replicas - replicas_to_run + 1, replicas + 1):
+            # Each replica can run with different arguments
+            simulator_args_current = _replace_callable_from_dict(
+                simulator_args, resource)
+            input_args_current = _replace_callable_from_dict(
+                input_args, resource)
+            print(f"Simulation arguments {simulator_args_current}")
+            # input_files either come from the template manager or
+            # are passed directly.
+            simulator_args_current["input_dir"] = input_files
+            simulator_args_current["on"] = resource
+
+            if input_args_current:
+                target_dir = ("benchmark_inputs/"
+                              f"{resource.machine_type}_{replica}")
```

Comment:
another weird line break.
Can we remove some identation? perhaps move some logic to a separate function?